### PR TITLE
Parse all Config's using their corresponding parser

### PR DESCRIPTION
**Pull request by teoljungberg**
- The Yaml parsers can return `nil`, so guard against such a case by
  returning an empty hash which will work cleanly further up.
- Replace `shared_example "a service based linter"`.

### CODE REVIEW

File: app/models/config/coffee_script.rb [hunk 1]
```diff
@@ -10,7 +10,7 @@ def linter_names
     private
 
     def parse(file_content)
-      result = Parser.json(file_content)
+      result = Config::Parser.json(file_content)
```
Comment (gylaz) [inline app/models/config/coffee_script.rb]: Why is it necessary to specify the namespace?

File: app/models/config/jscs.rb [hunk 2]
```diff
@@ -1,9 +1,21 @@
 module Config
   class Jscs < Base
+    def content
+      @content ||= super || default_content
```
Comment (gylaz) [inline app/models/config/jscs.rb]: Any reason to memoize here but on in `Eslint#content`?

File: app/models/config/python.rb [hunk 3]
```diff
@@ -1,9 +1,17 @@
 module Config
   class Python < Base
+    def content
+      @content ||= super.presence || default_content
```
Comment (gylaz) [inline app/models/config/python.rb]: Why use `.presence` here but in other `content` methods?

**Comment by gylaz** [comment]
:+1:

File: app/models/config/python.rb (hunk 3 above)
Comment (teoljungberg) [inline app/models/config/python.rb]: Parsing an empty string using the INI parser returns an empty hash, which is why I added this.

But, re-thinking it. I'll move it to the parser.

File: app/models/config/jscs.rb (hunk 2 above)
Comment (teoljungberg) [inline app/models/config/jscs.rb]: Good call, I think all of them should be memoized.

File: app/models/config/coffee_script.rb (hunk 1 above)
Comment (teoljungberg) [inline app/models/config/coffee_script.rb]: I experienced intermittent failures without this.

**Comment by teoljungberg** [comment]
I've tested this for

JSCS: https://github.com/teoljungberg/houndradish/pull/58
JSHint: https://github.com/teoljungberg/houndradish/pull/60
MDast: https://github.com/teoljungberg/houndradish/pull/61
Ruby: https://github.com/teoljungberg/houndradish/pull/62
Scss: https://github.com/teoljungberg/houndradish/pull/59
Python: https://github.com/teoljungberg/houndradish/pull/63

Which uses the different file formats that could have been effected. But none were.

File: app/models/config/coffee_script.rb (hunk 1 above)
Comment (gylaz) [inline app/models/config/coffee_script.rb]: Are you using Spring? I'd love to understand why we need to reference a namespace for code that is inside of that namespace. In theory, we shouldn't need to do that.

File: app/models/config/eslint.rb [hunk 4]
```diff
@@ -1,9 +1,17 @@
...
   class Eslint < Base
+    def serialize(data = content)
+      Config::Serializer.json(data)
+    end
+
     private
 
     def parse(file_content)
-      Parser.raw(file_content)
+      Config::Parser.yaml(file_content)
```
Comment (thorncp) [inline app/models/config/eslint.rb]: Should this be `json`? Or is this the Hound YAML config?

Comment (teoljungberg) [inline app/models/config/eslint.rb]: This is the config file for ESLint, so an `.eslintrc`. Which actually isn't valid JSON, but it is valid YAML. And YAML is a superset of JSON, which I had to learn the hard way this week.

File: app/models/config/coffee_script.rb (hunk 1 above)
Comment (teoljungberg) [inline app/models/config/coffee_script.rb]: I'm not. I'll see if I can reach some clarity to why I got a constant missmatch. I'll look into it

Comment (teoljungberg) [inline app/models/config/coffee_script.rb]: It seems to have been a constant-finding issue, requiring the parser class we want resolved it. 
Which now is fixed.

**Pull request merged**